### PR TITLE
add support for new dependencyResolutionManagement

### PR DESCRIPTION
**Pull request by Logiquo**

### CODE REVIEW

**Comment by Logiquo** [comment]
This is related to #395.

**Comment by modmuss50** [comment]
Had a quick look from my phone and looks good. 👍 can you add an integration test for this?

Let me know if you need help with that.

File: src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
Comment (liach) [inline src/main/java/net/fabricmc/loom/LoomGradlePlugin.java:50]: The diff can be simplified by retaining `apply(Project)` and adding a `apply(PluginAware)` at the same time. The override will move to the `apply(PluginAware)`, which will be:
```java
	@Override
	public void apply(PluginAware target) {
		target.getPlugins().apply(LoomRepositoryPlugin.class);
		if (target instanceof Project) {
			apply((Project) target);
		}
	}
```

File: src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java
Comment (liach) [inline src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: Mind rename to `getProjectUuid` which is the fabric style?

File: src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java
Comment (liach) [inline src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java:201]: So here, when the count is 0, I think the best approach is to leave a warning or fail the build (maybe add a command-line option to control that) when the desired maven doesn't exist. Imo we should fail by default, but a system property for warning can allow people to use proxy mavens.

File: src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomGradlePlugin.java]:
```suggestion
		if (target instanceof Project project) {
```

File: src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: This should also follow the naming conventions (`projectMappedClassifier`)

File: src/main/java/net/fabricmc/loom/configuration/MavenConfiguration.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/MavenConfiguration.java]: I think this can just be removed now - all it did was set up those repos.

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java]:
```suggestion
		} else if (target instanceof Project project) {
```

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java]:
```suggestion
		if (target instanceof Settings settings) {
```

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java]:
```suggestion
		if (target instanceof Settings settings) {
			gradleUserHomeDir = settings.getGradle().getGradleUserHomeDir();
		} else if (target instanceof Project project) {
			gradleUserHomeDir = project.getGradle().getGradleUserHomeDir();
```

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java]:
```suggestion
		if (target instanceof Settings settings) {
			rootDir = settings.getRootDir();
		} else if (target instanceof Project project) {
			rootDir = project.getRootDir();
```

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java]:
```suggestion
		if (target instanceof Settings settings) {
			rootDir = settings.getRootDir();
		} else if (target instanceof Project project) {
			rootDir = project.getRootDir();
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

**Comment by Logiquo** [comment]
> Had a quick look from my phone and looks good. 👍 can you add an integration test for this?
> 
> Let me know if you need help with that.

Hi, I did not have any experience in writing integration tests. If you can give me some examples/tutorials then I will be happy to write it.

**Comment by liach** [comment]
@LogicFan seems what you said in https://github.com/FabricMC/fabric-loom/issues/395#issuecomment-841046582 still needs to be addressed?

**Comment by Logiquo** [comment]
> @LogicFan seems what you said in [#395 (comment)](https://github.com/FabricMC/fabric-loom/issues/395#issuecomment-841046582) still needs to be addressed?

I agree with your solution, however, I am still testing this part of the code to make sure the behaviour is expected. So a commit will be up for today or tomorrow after I am confident that my code is correct.

**Comment by liach** [comment]
Also on unit tests, I think you will make a backbone project at https://github.com/FabricMC/fabric-loom/tree/dbc1f4cba7ad6c9bbce0aea46e682897a61c53dd/src/test/resources/projects and add a driver code file at https://github.com/FabricMC/fabric-loom/tree/dbc1f4cba7ad6c9bbce0aea46e682897a61c53dd/src/test/groovy/net/fabricmc/loom/test/integration

File: src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java
Comment (Logiquo) [reply src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java:201]: The `settings.getDependencyResolutionManagement().getRepositories()` and `project.getRepositories()` are different object in gradle. Since we do not add any repo into `project.getRepositories()` if `dependencyResolutionManagement` used, the `project.getRepositories()` will always be empty. Furthermore, there is currently no way of retrieving `settings` in project lifecycle, this means we do not know if a repo is added or not. Thus, the `count` will always be zero. 

This limitation will cause your idea hard to implement, and hence I prefer the existing code (I really don't have any other solution for this).

**Comment by Logiquo** [comment]
The test is added. Waiting for approval.

File: src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener
Comment (modmuss50) [inline src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener]: This isnt a valid accesswidener, I would remove it as I dont see how its realted to this test.

Comment (modmuss50) [reply src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener]: Oh, I see it forces a project mapped jar, still not a valid class name

Comment (Logiquo) [reply src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener]: Sorry, I do not have much experience with how the accesswidener works. This test is to force a build to use a project-mapped jar. Could you change this to a valid accesswidener? Thanks.

File: src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:100]: Might be out of scope for this PR, but it does duplicate some stuff in LoomGradleExtension, moving this out into its own class and removing the duplicates might not be a bad idea. 

At somepoint I want to give LGE a good cleanup so im not too fussed about doing this now.

**Review by modmuss50 — APPROVED** [review]

File: src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener
Comment (modmuss50) [inline src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener]:
```suggestion
extendable   class   net/minecraft/fluid/FluidState
```

**Comment by Logiquo** [comment]
Hi, may I ask what is the ETA to merge this PR? Is there anything more I need to do? This is my first PR in GitHub so I do not have much experience.

**Review by liach — APPROVED** [review]

**Comment by modmuss50** [comment]
Should be ready to merge, I just want to test it locally first.

File: src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: Im not sure if a whole UUID is needed? would just the project name (formatted in a classifier safe format) be a better solution?

Comment (Logiquo) [reply src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: In theory, anything that unique for the project should be fine. UUID is just for naming collision avoidance. I don't know if the project name would be safe enough for this. Can people define two sub-projects with the same name?

Comment (liach) [reply src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: I think if you use full subproject names, you are fine. You can define like `:sub1:duplicate` and `:sub2:duplicate` subprojects within one project I think, but their full project names differ.

Comment (Logiquo) [reply src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: I see, I will change it to the fully-qualified-name. It will be more human-readable, and hopefully shorter. A commit will be up for this tomorrow.

Comment (Logiquo) [reply src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java]: A commit is up. My biggest concern is what if there is invalid char (i.e. cannot exists in the path) inside the project name. Any solution for this? Should we write a transformer that replaces all the invalid char?

I cannot find a good resource for what is a valid name for the Gradle project, seems and char except `:` are legal.

**Pull request merged**
